File: leetcode/python/110_balanced_binary_tree.py
"""
T: O(N)
S: O(N)

Simple recursive solution based on determining the depth of subtrees.
If some subtree is inbalanced - mark its depth as such and propagate it.
"""

from typing import Dict, Optional


class TreeNode:

    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def isBalanced(self, root: TreeNode) -> bool:
        node_depth = {None: 0}  # type: Dict[Optional[TreeNode], int]
        nodes_to_process = [(root, False)]
        while nodes_to_process:
            curr, is_processed = nodes_to_process.pop()
            if not curr:
                continue
            if is_processed:
                left = node_depth[curr.left]
                right = node_depth[curr.right]
                if abs(left - right) > 1:
                    return False
                node_depth[curr] = 1 + max(left, right)
            else:
                nodes_to_process.append((curr, True))
                nodes_to_process.append((curr.left, False))
                nodes_to_process.append((curr.right, False))
        return True

File: leetcode/java/110BalancedBinaryTree.java
/**
 * T: O(N) S: O(N)
 *
 * <p>Traverse the tree, get depth of left and right children. If the current subtree is unbalanced
 * or left or right subtree is unbalanced - mark current subtree as unbalanced. Otherwise, record
 * the real maximum depth of the current subtree - depth of the deeper child(left or right) plus
 * one. This way, if any subtree is unbalanced - this information will be propagated to the root.
 */
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
}

class Solution {
  private final int DEPTH_OF_UNBALANCED_TREE = -1;

  public boolean isBalanced(TreeNode root) {
    Map<TreeNode, Integer> nodeDepths = new HashMap<>();
    nodeDepths.put(null, 0);
    Stack<TreeNode> nodesToProcess = new Stack<>();
    nodesToProcess.push(root);
    while (!nodesToProcess.isEmpty()) {
      TreeNode currentNode = nodesToProcess.pop();
      if (currentNode == null) {
        continue;
      }
      TreeNode left = currentNode.left;
      TreeNode right = currentNode.right;
      if (nodeDepths.containsKey(left) && nodeDepths.containsKey(right)) {
        int leftDepth = nodeDepths.get(left);
        int rightDepth = nodeDepths.get(right);
        if (leftDepth == DEPTH_OF_UNBALANCED_TREE
            || rightDepth == DEPTH_OF_UNBALANCED_TREE
            || Math.abs(leftDepth - rightDepth) > 1) {
          nodeDepths.put(currentNode, DEPTH_OF_UNBALANCED_TREE);
        } else {
          nodeDepths.put(currentNode, Math.max(leftDepth, rightDepth) + 1);
        }
      } else {
        nodesToProcess.push(currentNode);
        nodesToProcess.push(right);
        nodesToProcess.push(left);
      }
    }
    return nodeDepths.get(root) != DEPTH_OF_UNBALANCED_TREE;
  }
}
